internal/serialization/default_compact_writer.go

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package serialization
+
+import (
+	"fmt"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/check"
+	ihzerrors "github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
+	pserialization "github.com/hazelcast/hazelcast-go-client/serialization"
+)
+
+type DefaultCompactWriter struct {
+	out               *PositionalObjectDataOutput
+	serializer        CompactStreamSerializer
+	fieldOffsets      []int32
+	schema            Schema
+	dataStartPosition int32
+}
+
+func NewDefaultCompactWriter(serializer CompactStreamSerializer, out *PositionalObjectDataOutput, schema Schema) DefaultCompactWriter {
+	var fieldOffsets []int32
+	var dataStartPosition int32
+	if schema.numberOfVarSizeFields != 0 {
+		fieldOffsets = make([]int32, schema.numberOfVarSizeFields)
+		dataStartPosition = out.Position() + Int32SizeInBytes
+		// Skip for length and primitives.
+		out.WriteZeroBytes(int(schema.fixedSizeFieldsLength + Int32SizeInBytes))
+	} else {
+		dataStartPosition = out.Position()
+		// Skip for primitives. No need to write data length, when there is no
+		// variable-size fields.
+		out.WriteZeroBytes(int(schema.fixedSizeFieldsLength))
+	}
+	return DefaultCompactWriter{
+		serializer:        serializer,
+		out:               out,
+		schema:            schema,
+		fieldOffsets:      fieldOffsets,
+		dataStartPosition: dataStartPosition,
+	}
+}
+
+func (r DefaultCompactWriter) WriteInt32(fieldName string, value int32) {
+	position := r.getFixedSizeFieldPosition(fieldName, pserialization.FieldKindInt32)
+	r.out.PWriteInt32(position, value)
+}
+
+func (r DefaultCompactWriter) WriteString(fieldName string, value *string) {
+	r.writeVariableSizeField(fieldName, pserialization.FieldKindString, value, func(out *PositionalObjectDataOutput, v interface{}) {
+		out.WriteString(*v.(*string))
```

Comment:
Should there be a check for nil string?

Reply:
This reader method is not called if value is nil. There is a check in `writeVariableSizeField` method